huggan/pytorch/sngan/train.py

```diff
@@ -0,0 +1,371 @@
+
+            # train with fake
+            noise = torch.randn(batch_size, args.latent_dim, 1, 1, device=accelerator.device)
+            fake = generator(noise)
+            label.fill_(fake_label)
+            output = discriminator(fake.detach())
+            loss_d_fake = criterion(output, label)
+            accelerator.backward(loss_d_fake)
+            discriminator_optimizer.step()
+
+            avg_loss_d.update((loss_d_real + loss_d_fake).item(), batch_size)
+            avg_d_g_z1.update(output.mean().item())
+
+            ############################
+            # (2) Update G model: maximize log(D(G(z)))
+            ###########################
+            generator.zero_grad()
+            label.fill_(real_label)  # fake labels are real for generator cost
+            output = discriminator(fake)
+            # minimize loss but also maximize alpha channel
+            loss_g = criterion(output, label) + fake[:, -1].mean()
+            accelerator.backward(loss_g)
+            generator_optimizer.step()
+
+            avg_loss_g.update(loss_g.item(), batch_size)
+            avg_d_g_z2.update(output.mean().item())
+
+        # write logs
+        with open(f"{args.output_dir}/logs.csv", "a") as f:
+            csv.writer(f).writerow(
+                [epoch, avg_loss_g, avg_loss_d, avg_d_x, avg_d_g_z1, avg_d_g_z2]
+            )
+
+        train_logs = {
+            "epoch": epoch,
+            "discriminator_loss": avg_loss_d.avg,
+            "generator_loss": avg_loss_g.avg,
+            "D_x": avg_d_x.avg,
+            "D_G_z1": avg_d_g_z1.avg,
+            "D_G_z2": avg_d_g_z2.avg,
+        }
+        if accelerator.is_local_main_process:
+            if args.wandb:
+                wandb.log(train_logs)
+
+        if (epoch + 1) % args.logging_steps == 0:
+            # save samples
+            fake = generator(fixed_noise)
+            file_name = f"{args.output_dir}/fake_samples_epoch_{epoch}.png"
+            vutils.save_image(
+                fake.detach(),
+                file_name,
+                normalize=True,
+            )
+
+            if accelerator.is_local_main_process and args.wandb:
+                wandb.log({'generated_examples': wandb.Image(str(file_name))})
+
+            # save_checkpoints
+            torch.save(generator.state_dict(), f"{args.output_dir}/generator_epoch_{epoch}.pth")
+            torch.save(discriminator.state_dict(), f"{args.output_dir}/discriminator_epoch_{epoch}.pth")
+
+            # Calculate FID metric
+            fid = calculate_fretchet(real_cpu, fake, model.to(accelerator.device))
+            logger.info(f"FID: {fid}")
+            if accelerator.is_local_main_process and args.wandb:
+                wandb.log({"FID": fid})
+
+    # Optionally push to hub
+    if accelerator.is_main_process and args.push_to_hub:
+        # os.makedirs(f"{args.output_dir}/{args.model_name}", exist_ok=True)
+        generator.push_to_hub(
```

Comment:
If this doesn't work reliably, see:

Here:
https://github.com/huggingface/community-events/blob/15c92b6375d12febb9b8bde8b3abba70b429929d/huggan/pytorch/pix2pix/train.py#L115-L117

And Here:
https://github.com/huggingface/community-events/blob/15c92b6375d12febb9b8bde8b3abba70b429929d/huggan/pytorch/pix2pix/train.py#L292-L297

Reply:
It works ok now. Just forgot to remove the comment